implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+            state: HandshakeState::new(static_key, ephemeral_key, message1_payload),
+        })
+    }
+
+    /// Import the k secret
+    async fn import_k_secret(&self, content: Vec<u8>) -> Result<KeyId> {
+        self.vault
+            .import_ephemeral_secret(Secret::new(content), Self::k_attributes())
+            .await
+    }
+
+    /// Import the ck secret
+    async fn import_ck_secret(&self, content: Vec<u8>) -> Result<KeyId> {
+        self.vault
+            .import_ephemeral_secret(Secret::new(content), Self::ck_attributes())
+            .await
+    }
+
+    /// Return the public key corresponding to a given key id
+    async fn get_public_key(&self, key_id: &KeyId) -> Result<PublicKey> {
+        self.vault.get_public_key(key_id).await
+    }
+
+    /// Compute a Diffie-Hellman key between a given key id and the other party public key
+    async fn dh(&self, key_id: &KeyId, public_key: &PublicKey) -> Result<KeyId> {
+        self.vault.ec_diffie_hellman(key_id, public_key).await
+    }
+
+    /// Compute two derived ck, and k keys based on existing ck and k keys + an optional
+    /// Diffie-Hellman key
+    async fn hkdf(&self, state: &mut HandshakeState, dh: Option<&KeyId>) -> Result<(KeyId, KeyId)> {
+        // if dh is not defined that means that we are doing the last step in the noise protocol and splitting
+        // an encryption key and a decryption key. In that case this means that we produce AES keys for both
+        // keys
+        let ck_attributes = if dh.is_none() {
+            Self::k_attributes()
+        } else {
+            Self::ck_attributes()
+        };
+
+        let hkdf_output = self
+            .vault
+            .hkdf_sha256(
+                &state.ck,
+                b"",
+                dh,
+                vec![ck_attributes, Self::k_attributes()],
+            )
+            .await?;
+
+        match hkdf_output.as_slice() {
+            [new_ck, new_k] => {
+                self.vault.delete_ephemeral_secret(state.ck.clone()).await?;
+                self.vault.delete_ephemeral_secret(state.k.clone()).await?;
+                state.n = 0;
+                Ok((new_ck.into(), new_k.into()))
+            }
+            _ => Err(XXError::InternalVaultError.into()),
+        }
+    }
+
+    /// Decrypt a ciphertext 'c' using the key 'k' and the additional data 'h'
+    async fn hash_and_decrypt(&self, state: &mut HandshakeState, c: &[u8]) -> Result<Vec<u8>> {
+        let mut nonce = [0u8; 12];
+        nonce[10..].copy_from_slice(&state.n.to_be_bytes());
+        let result = self
+            .vault
+            .aead_aes_gcm_decrypt(&state.k, c, nonce.as_ref(), &state.h)
+            .await
+            .map(|b| b.to_vec())?;
+        state.mix_hash(c);
+        state.n += 1;
+        Ok(result)
+    }
+
+    /// Encrypt a plaintext 'c' using the key 'k' and the additional data 'h'
+    async fn encrypt_and_hash(&self, state: &mut HandshakeState, c: &[u8]) -> Result<Vec<u8>> {
```

Comment:
Can we please have different names for encrypted and decrypted binaries? I guess `c` is cipher text, then maybe plain text could be `p`?

Reply:
Yes, I adjusted the naming